connect/runtime/src/main/java/org/apache/kafka/connect/runtime/RestartPlan.java

```diff
@@ -0,0 +1,154 @@
+
+    /**
+     * Create a new request to restart a connector and optionally its tasks.
+     *
+     * @param request          the restart request; may not be null
+     * @param restartStateInfo the current state info for the connector; may not be null
+     */
+    public RestartPlan(RestartRequest request, ConnectorStateInfo restartStateInfo) {
+        this.request = Objects.requireNonNull(request, "RestartRequest name may not be null");
+        this.stateInfo = Objects.requireNonNull(restartStateInfo, "ConnectorStateInfo name may not be null");
+        // Collect the task IDs to stop and restart (may be none)
+        idsToRestart = Collections.unmodifiableList(
+                stateInfo.tasks()
+                        .stream()
+                        .filter(this::isRestarting)
+                        .map(taskState -> new ConnectorTaskId(request.connectorName(), taskState.id()))
+                        .collect(Collectors.toList())
+        );
+    }
+
+    /**
+     * Get the connector name.
+     *
+     * @return the name of the connector; never null
+     */
+    public String connectorName() {
+        return request.connectorName();
+    }
+
+    /**
+     * Get the original {@link RestartRequest}.
+     *
+     * @return the restart request; never null
+     */
+    public RestartRequest restartRequest() {
+        return request;
+    }
+
+    /**
+     * Get the {@link ConnectorStateInfo} that reflects the current state of the connector <em>except</em> with the {@code status}
+     * set to {@link AbstractStatus.State#RESTARTING} for the {@link Connector} instance and any {@link Task} instances that
+     * are to be restarted, based upon the {@link #restartRequest() restart request}.
+     *
+     * @return the connector state info that reflects the restart plan; never null
+     */
+    public ConnectorStateInfo restartConnectorStateInfo() {
+        return stateInfo;
+    }
+
+    /**
+     * Get the immutable collection of {@link ConnectorTaskId} for all tasks to be restarted
+     * based upon the {@link #restartRequest() restart request}.
+     *
+     * @return the IDs of the tasks to be restarted; never null but possibly empty
+     */
+    public Collection<ConnectorTaskId> taskIdsToRestart() {
+        return idsToRestart;
+    }
+
+    /**
+     * Determine whether the {@link Connector} instance is to be restarted
+     * based upon the {@link #restartRequest() restart request}.
+     *
+     * @return true if the {@link Connector} instance is to be restarted, or false otherwise
+     */
+    public boolean restartConnector() {
+        return isRestarting(stateInfo.connector());
+    }
+
+    /**
+     * Determine whether at least one {@link Task} instance is to be restarted
+     * based upon the {@link #restartRequest() restart request}.
+     *
+     * @return true if any {@link Task} instances are to be restarted, or false if none are to be restarted
+     */
+    public boolean restartAnyTasks() {
```

Comment:
@kkonstantine  and @rhauch  I renamed the methods for shouldXXX suggestion.

For removing Any in the name, I think we should keep it in the name because plan.shouldRestartTasks() in caller code conveys the intent it would restart all tasks but plan.shouldRestartAnyTask conveys it would restart at least one task. I know its again not the best name.

wdyt if we rename it to shouldRestartAtleastOneTask, it was too verbose that's why we had picked Any?

Reply:
Renamed to shouldRestartTasks